File: src/ssh.ts
import { Client, type ConnectConfig, type SFTPWrapper } from "ssh2";
import { readFileSync } from "fs";
import type { ServerEntry } from "./types.js";

const connections = new Map<string, Client>();

function buildConnectConfig(server: ServerEntry): ConnectConfig {
  const config: ConnectConfig = {
    host: server.host,
    port: server.port,
    username: server.username,
    readyTimeout: 10000,
  };

  if (server.authMethod === "password" && server.password) {
    config.password = server.password;
  } else if (server.authMethod === "private_key") {
    if (server.privateKeyPath) {
      config.privateKey = readFileSync(server.privateKeyPath, "utf-8");
    } else if (server.privateKey) {
      config.privateKey = server.privateKey;
    }
  }

  return config;
}

export async function getConnection(server: ServerEntry): Promise<Client> {
  // Check pool for existing connection
  const existing = connections.get(server.id);
  if (existing) {
    try {
      await new Promise<void>((resolve, reject) => {
        existing.exec("true", (err, stream) => {
          if (err) return reject(err);
          stream.on("close", () => resolve()).on("error", reject);
        });
      });
      return existing;
    } catch {
      connections.delete(server.id);
      existing.end();
    }
  }

  return new Promise((resolve, reject) => {
    const client = new Client();
    client.on("ready", () => {
      connections.set(server.id, client);
      resolve(client);
    });
    client.on("error", (err) => {
      reject(new Error(`SSH connection to ${server.host}:${server.port} failed: ${err.message}`));
    });
    client.connect(buildConnectConfig(server));
  });
}

export function closeConnection(serverId: string): void {
  const conn = connections.get(serverId);
  if (conn) {
    conn.end();
    connections.delete(serverId);
  }
}

export function closeAllConnections(): void {
  for (const [id, conn] of connections) {
    conn.end();
    connections.delete(id);
  }
}

export async function execCommand(
  client: Client,
  command: string,
  timeoutMs: number = 30000
): Promise<{ exitCode: number; stdout: string; stderr: string }> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      reject(new Error(`Command timed out after ${timeoutMs}ms`));
    }, timeoutMs);

    client.exec(command, (err, stream) => {
      if (err) {
        clearTimeout(timer);
        return reject(err);
      }

      let stdout = "";
      let stderr = "";

      stream.on("close", (code: number | null) => {
        clearTimeout(timer);
        resolve({ exitCode: code ?? 0, stdout, stderr });
      });

      stream.on("data", (data: Buffer) => {
        stdout += data.toString();
      });

      stream.stderr?.on("data", (data: Buffer) => {
        stderr += data.toString();
      });
    });
  });
}

export async function uploadFile(
  client: Client,
  localPath: string,
  remotePath: string
): Promise<void> {
  return new Promise((resolve, reject) => {
    client.sftp((err, sftp: SFTPWrapper) => {
      if (err) return reject(err);
      const readStream = require("fs").createReadStream(localPath);
      const writeStream = sftp.createWriteStream(remotePath);
      writeStream.on("close", () => { sftp.end(); resolve(); });
      writeStream.on("error", (err: Error) => { sftp.end(); reject(err); });
      readStream.pipe(writeStream);
    });
  });
}

export async function uploadContent(
  client: Client,
  content: string,
  remotePath: string
): Promise<void> {
  return new Promise((resolve, reject) => {
    client.sftp((err, sftp: SFTPWrapper) => {
      if (err) return reject(err);
      const writeStream = sftp.createWriteStream(remotePath);
      writeStream.on("close", () => { sftp.end(); resolve(); });
      writeStream.on("error", (err: Error) => { sftp.end(); reject(err); });
      writeStream.write(content);
      writeStream.end();
    });
  });
}

File: python/src/mcp_deploy_server/ssh.py
"""Async SSH connection manager with connection pooling per server."""

import asyncio
from pathlib import Path
from typing import Any

import asyncssh

from .models import ServerEntry

_connections: dict[str, asyncssh.SSHClientConnection] = {}


def _build_connect_kwargs(server: ServerEntry) -> dict[str, Any]:
    kwargs: dict[str, Any] = {
        "host": server.host,
        "port": server.port,
        "username": server.username,
        "known_hosts": None,
        "connect_timeout": 10,
    }
    if server.auth_method == "password" and server.password:
        kwargs["password"] = server.password
    elif server.auth_method == "private_key":
        if server.private_key_path:
            kwargs["client_keys"] = [server.private_key_path]
        elif server.private_key:
            kwargs["client_keys"] = [(asyncssh.import_private_key(server.private_key))]
    return kwargs


async def get_connection(server: ServerEntry) -> asyncssh.SSHClientConnection:
    """Get or create a pooled SSH connection for a server."""
    conn = _connections.get(server.id)
    if conn is not None:
        try:
            # Liveness check
            result = await conn.run("true", check=False)
            if result.exit_code == 0:
                return conn
        except Exception:
            pass
        # Stale connection, remove
        _connections.pop(server.id, None)
        try:
            conn.close()
        except Exception:
            pass

    kwargs = _build_connect_kwargs(server)
    conn = await asyncssh.connect(**kwargs)
    _connections[server.id] = conn
    return conn


def close_connection(server_id: str) -> None:
    """Close a specific server connection."""
    conn = _connections.pop(server_id, None)
    if conn:
        conn.close()


def close_all_connections() -> None:
    """Close all pooled connections."""
    for conn in _connections.values():
        conn.close()
    _connections.clear()


async def exec_command(
    conn: asyncssh.SSHClientConnection,
    command: str,
    timeout_sec: int = 30,
) -> dict[str, Any]:
    """Execute a command and return exit_code, stdout, stderr."""
    try:
        result = await asyncio.wait_for(
            conn.run(command, check=False),
            timeout=timeout_sec,
        )
        return {
            "exit_code": getattr(result, "exit_status", 0) or 0,
            "stdout": result.stdout or "",
            "stderr": result.stderr or "",
        }
    except asyncio.TimeoutError:
        raise TimeoutError(f"Command timed out after {timeout_sec}s")


async def upload_content(
    conn: asyncssh.SSHClientConnection,
    content: str,
    remote_path: str,
) -> None:
    """Write content to a remote file via SFTP."""
    async with conn.start_sftp_client() as sftp:
        async with sftp.open(remote_path, "w") as f:
            await f.write(content)


async def upload_file(
    conn: asyncssh.SSHClientConnection,
    local_path: str,
    remote_path: str,
) -> None:
    """Upload a local file to a remote path via SFTP."""
    async with conn.start_sftp_client() as sftp:
        await sftp.put(local_path, remote_path)
